### ISSUE TITLE
Docker engine does not start if Swarm --listen-addr IP not available at startup

### ISSUE THREAD

**Issue by tfasz** [moby/moby#30190]
**Description**

I have a 2 node Swarm cluster running across a VPN connection. The VPN connection is established and managed by a docker container running on Host 1. I can successfully start up the VPN container on Host 1 and init the Swarm cluster but as soon as Host 1 reboots Docker will not start.

**Steps to reproduce the issue:**
1. Start VPN container on Host 1 - establish tunnel between Host 1 and Host 2 with VPN IPs 192.168.250.1 and 192.168.250.2 respectively. 
2. Init swarm cluster on Host 1 and join swarm cluster from Host 2.
3. Reboot Host 1.

**Describe the results you received:**

Docker engine fails to start with below error.
```
Jan 16 19:19:42 local-dev docker[853]: time="2017-01-16T19:19:42.565279567Z" level=info msg="Loading containers: done."
Jan 16 19:19:42 local-dev docker[853]: time="2017-01-16T19:19:42.616559922Z" level=error msg="cluster exited with error: failed to listen on remote API address: listen tcp 192.168.250.1:2377: bind: cannot assign requested address"
Jan 16 19:19:42 local-dev docker[853]: time="2017-01-16T19:19:42.616591134Z" level=fatal msg="Error creating cluster component: swarm component could not be started: failed to listen on remote API address: listen tcp 192.168.250.1:2377: bind: cannot assign requested address"
Jan 16 19:19:42 local-dev systemd[1]: docker.service: Main process exited, code=exited, status=1/FAILURE
Jan 16 19:19:42 local-dev systemd[1]: Failed to start Docker Application Container Engine.
Jan 16 19:19:42 local-dev systemd[1]: docker.service: Unit entered failed state.
Jan 16 19:19:42 local-dev systemd[1]: docker.service: Failed with result 'exit-code'.
```

**Describe the results you expected:**

It does not seem like most users would expect initialization of the Swarm cluster to prevent Docker Engine from starting successfully. Docker engine should successfully start and join the Swarm cluster asynchronously in background in a retry loop until successful. This would allow all local Docker containers to start including any container that is responsible for setting up Swarm network.

**Additional information you deem important (e.g. issue happens only occasionally):**

Testing this with Docker 1.13 but guessing it will also happen with 1.12. The VPN container uses [TINC VPN](http://tinc-vpn.org/) and is run with a compose file:
```
version: '2'
services:
    vmmgr:
        image: vmmgrimage
        volumes:
            - /dev/net/tun:/dev/net/tun
        cap_add:
            - NET_ADMIN
        network_mode: "host"
```

The swarm cluster is started once the VPN connection is established with:
```
docker swarm init --advertise-addr 192.168.250.1:2377 --listen-addr 192.168.250.1:2377
TOKEN=$(docker swarm join-token -q manager)
docker -H 192.168.250.2:2375 swarm join --advertise-addr 192.168.250.2:2377 --listen-addr 192.168.250.2:2377 --token ${TOKEN} 192.168.250.1:2377
```

**Output of `docker version`:**

```
Client:
 Version:      1.13.0-rc6
 API version:  1.25
 Go version:   go1.7.3
 Git commit:   2f2d055
 Built:        Wed Jan 11 18:17:53 2017
 OS/Arch:      linux/amd64

Server:
 Version:      1.13.0-rc6
 API version:  1.25 (minimum version 1.12)
 Go version:   go1.7.3
 Git commit:   2f2d055
 Built:        Wed Jan 11 18:17:53 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.13.0-rc6
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 03e5862ec0d8d3b3f750e19fca3ee367e13c090e
runc version: 2f7393a47307a16f8cee44a37b262e8b81021e3e
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 11.46 GiB
Name: local-dev
ID: OKE6:NUAH:LYOY:FP4T:GWAG:AB3P:HQGE:GL2H:BKSD:KBZP:X7GV:SZDL
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Host 1 is a VirtualBox VM running locally and Host 2 is a EC2 instance. I'd like to use a Docker container to establish/manage the remote EC2 instance + VPN so that I can use Docker for Mac or Docker for Windows in the future to run the local Host 1 and connect it to a remote EC2 Host 2.

**Comment by fntlnz** [comment]
What you're trying to do is *very interesting* but have a few limitations because:

- You are expecting a consistent behavior in an environment where is not possible to achieve real consistency by design (only 2 nodes)
- dockerd inheritly depends on the network stack thus mean that you would need another process (or docker daemon) which creates the VPN in advance and then do your stuff

The only possible way I see now to achieve this (in the same daemon, using containers) is to try to do this using the new docker engine managed (as containers) plugins coming in 1.13 (see: https://github.com/docker/docker/blob/48a9e53d70472bebad908b273351d8a07939a764/docs/extend/index.md)

**Comment by CrimsonGlory** [comment]
Seems related to #29580

**Comment by CrimsonGlory** [comment]
Seems to be fixed in #29580 via #31631.
